Clean up Machine.java so that it is less of a data class.

smit9025 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
I took Machine.java and removed the non-essential getters and setters. I extracted some math in changeState() and put it into machine.java so that machina.java handles it instead of machineshopsimulator.java. I then took the methods that were created in machine.java and combined them into another method that replaces the three lines in changeState() where the machine starts on the next job. I also replaced the use of machine[themachine] with currentmachine by setting machine[themachine] to the variable currentmachine.
## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- Please link to the issue here: -->
close #4 
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- Describe the smell this addresses and the particular location the smell occurs -->
<!--- Provide a summary of the description in the issue, but don’t necessarily copy all the details over to the pull request. -->
<!--- Make sure to include enough information that people don’t have to click over the issue if they aren’t inclined. -->
<!--- You need to make sure the issue has the necessary detail. -->
This change is required because it reduces the data class smell by adding more functionality to a class. It also increases readability.
The data class smell is located in job.java, machine.java, task.java, simulationspecification.java, simulationresults.java, jobspecification.java, and jobcompletiondata.java.
The description in the issues basically says to clean machine.java so that it isn't a data class anymore and has more use.
## Smell impact
<!--- What smells does this reduce/eliminate? -->
<!--- What smells does this introduce/increase? -->
<!--- Were there old smells that you hadn’t noticed that this helps bring to your attention? -->
This reduces the data class smell. It doesn't introduce/increase any other smell as far as I can tell. I didn't discover any new smells.
## Testing & Screenshots:
<!--- Screenshot of the code coverage report so we can track that over time. -->
<!--- Mention any new tests here as well. -->
<img width="947" alt="coverage" src="https://user-images.githubusercontent.com/46760134/93643520-c22faf80-f9c5-11ea-8b16-3cb380419654.PNG">

No new tests were made.
- [x] All new and existing tests passed.

smit9025 committed:
Refactor Machine.java so that it isn't as much of a data class anymore, and more of a constructor.
I refactored Machine.java by getting rid of getters and setters and just setting the available variables that can have data stored in them. This makes the class more of just a constructor.
Also changed string variables on top to be package-private instead of public.
Also changed methods in MachineShopSimulator.java to be private.

smit9025 committed:
Refactor Machine.java so that it isn't as much of a data class anymore, and more of a constructor.
I refactored Machine.java by getting rid of getters and setters and just setting the available variables that can have data stored in them. This makes the class more of just a constructor.
Also changed string variables on top to be package-private instead of public.
Also changed methods in MachineShopSimulator.java to be private.
Simplified a piece of the method changeState in MachineShopSimulator so that it looks better.

UtkarshKr007 reviewed (approved):
The code changes are easy to read and solves the issue you tackled. All the tests are passing. Great work!!

The commit messages somewhat confused me, because they seem to be the same except the second commit message had an extra line at the end describing the change in that commit. I would say just having that last line would suffice since I am only concerned about changes in that specific commit when looking at its commit message.

deglm006 reviewed (changes requested):
To me the motivation behind this is a bit unclear.
I feel like this might not be a great way to refactor the code. This seems to worsen inappropriate intimacy between Machine and MSS. I believe that the methods removed from Machine should remain, as if its implementation is updated (i.e. the jobs in Machine are represented by a different data structure) then only Machine.java would need to be updated as its interface would stay the same.
I don't feel like I can approve this without a stronger/clearer motivation behind it.

Also, I feel like the commit messages are a bit uninformative and It feels a bit weird to have two identical commit messages. The title of the PR I think could use some work.

For the code coverage, I believe the intent was to have both before and after, so that we can see if the coverage has increased or decreased.

deglm006 commented on src/applications/Machine.java:
```diff
@@ -4,50 +4,14 @@
-    private int changeTime; // machine change-over time
-    private int totalWait; // total delay at this machine
-    private int numTasks; // number of tasks processed on this machine
-    private Job activeJob; // job currently active on this machine
```
This seems like a bad thing. In general I think we want to try to make as many things private as possible.

deglm006 commented on src/applications/MachineShopSimulator.java:
```diff
@@ -47,38 +47,37 @@ boolean moveToNextMachine(Job theJob, SimulationResults simulationResults) {
-                        - machine[theMachine].getActiveJob().getArrivalTime());
-                machine[theMachine].setNumTasks(machine[theMachine].getNumTasks() + 1);
-                int t = machine[theMachine].getActiveJob().removeNextTask();
+                machine[theMachine].activeJob = ((Job) machine[theMachine].jobQ.remove());
+                machine[theMachine].totalWait =
+                        (machine[theMachine].totalWait + timeNow - machine[theMachine].activeJob.getArrivalTime());
+                machine[theMachine].numTasks++;
+                int t = machine[theMachine].activeJob.removeNextTask();
```
This chunk of code is doing a lot of work I think would be better handled in Machine. 
I think the "data class" smell may have been reduced, but I think now it only stores data and is kind of a lazy class which is worse imo.
I feel like to resolve the data class you went the wrong direction by pulling stuff into MSS, when it might have been better to pull logic into Machine.

NicMcPhee replied:
I think this comment is spot on. I think you want to be moving logic _into_ `Machine`, not pulling it out of it.

smit9025 committed:
Following review of the first pull request for this branch, I decided to make a change where I would move the changeState method to machine.java.
What I did was used refactoring tools to first make the method static, then I had the tools move the method from MachineShopSimulator.java to Machine.java.
I have also remade the initialized variables on top of Machine.java to be private along with required getters and a setter by encapsulation. I would like to be able to remove all getters and setters, but I'm not sure how yet.
I created two more methods for use in changeState that improve readability of the method.
All tests pass.

NicMcPhee commented on .idea/misc.xml:
```diff
@@ -3,7 +3,7 @@
     <option name="languageLevel" value="ES6" />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="false" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="false" project-jdk-name="11" project-jdk-type="JavaSDK">
```
You want to be a little careful about including things like this in a pull request that's arguably about something entirely different. In this case (because there's just the two of you), bumping up from Java 8 to 11 isn't a huge deal, but in general you'd want that to be a thing that is discussed and done as its own change, with documentation on _why_ you'd want to move to 11, etc., etc.

smit9025 replied:
That was an accidental change that happened because I only have Java 11, I will change that back.

NicMcPhee commented on src/applications/MachineShopSimulator.java:
```diff
@@ -4,11 +4,11 @@
-    public static final String CHANGE_OVER_TIME_MUST_BE_AT_LEAST_0 = "change-over time must be >= 0";
-    public static final String EACH_JOB_MUST_HAVE_AT_LEAST_1_TASK = "each job must have >= 1 task";
-    public static final String BAD_MACHINE_NUMBER_OR_TASK_TIME = "bad machine number or task time";
+    static final String NUMBER_OF_MACHINES_MUST_BE_AT_LEAST_1 = "number of machines must be >= 1";
```
Straight up question: Why are you changing the visibility of these? Why/how does that relate to making `Machine` less of a cata class?

smit9025 replied:
Me mindlessly accepting IntelliJ's suggestions. Has been reverted.

NicMcPhee commented on src/applications/MachineShopSimulator.java:
```diff
@@ -47,38 +47,37 @@ boolean moveToNextMachine(Job theJob, SimulationResults simulationResults) {
                                             // schedule next one.
         Job lastJob;
-        if (machine[theMachine].getActiveJob() == null) {// in idle or change-over
+        if (machine[theMachine].activeJob == null) {// in idle or change-over
```
This (and the change below on line 57) are only possible because you made those fields visible up above. Like @deglm006, I'm not convinced that making them more visible is a great move, which then affects these changes.

smit9025 replied:
This is very fair and has been reverted with the commit that has changes based on Joe's suggestions.

NicMcPhee reviewed (changes requested):
I agree with @deglm006 that this seems to be moving the wrong direction. This seems to be moving all the "interesting" stuff out of `Machine`, when I'd encourage you to be moving more logic _into_ machine.

Certainly dropping `private` from all those field in `Machine` seems like a smell to me, where you're losing encapsulation and moving more towards a Lazy Class.

smit9025 committed:
Resolving suggestions from Nic
Changed project-jdk-name="11" in misc.xml back to project-jdk-name="1.8"
Changed static string names in MachineShopSimulator.java back to public because it is true that doing so in the first place doesn't deal with Machine.java whatsoever. I'm pretty sure it happened because IntelliJ made a suggestion and I was like "Sounds good to me."

smit9025 committed:
Changed changeState from static by using refactoring tools "Convert to Instance Method"

deglm006 commented on src/applications/Machine.java:
```diff
@@ -15,39 +15,57 @@
+     *
+     * @return last job run on this machine
+     */
+    Job changeState(Machine[] machine, EventList eList, int largeTime, int timeNow, int theMachine) {// Task on theMachine has finished,
```
I feel like this is a lot of stuff to be passing to a method. This smells of long parameter length.

smit9025 replied:
That is what I also thought as well for being an issue. I'm not sure how to reduce the number of parameters, so I just moved the method back to machineshopsimulator.java

deglm006 reviewed (commented):
I feel like this is better than when I first reviewed. 
It does smell a bit of long parameter list and inappropriate intimacy, though.
I would suggest leaving changeState in MSS and move the most relevant parts into Machine.

smit9025 committed:
Reverted moving changeState method from MachineShopSimulator.java to machine.java. Now there are only two new functions in machine.java that were pulled from changeState that was math. I don't know what else to pull into machine.java

smit9025 committed:
Added a variable to changeState so that the index position of machine is stored and then the variable goes through the method instead of machine[i]. I then moved three lines into machine.java, these including the two new functions made in machine.java and the task increment up.

smit9025 committed:
Changed method testMethod to beginNextJob

deglm006 reviewed (approved):
Code looks good. The description of the PR definitely needs to be updated though.

deglm006 commented on src/applications/Machine.java:
```diff
@@ -15,10 +15,27 @@
         return jobQ;
     }
 
+    public void setJobQ(LinkedQueue jobQ) {
```
Is this method used for anything?

smit9025 replied:
It is not and will be removed promptly.

smit9025 committed:
Removed unused methods in Machine.java
